flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1264,6 +1278,326 @@ def test_all_transitions(self, desired_state, discovered_state):
+                now=datetime.now(tz=UTC),
+                dataset_id=local_dataset.dataset_id,
+                node_id=self.deployer.node_uuid,
+            )
+        )
+
+    def test_lease_elsewhere(self):
+        """
+        If there is a lease for a dataset on another node, there isn't a
+        corresponding desired dataset.
+        """
+        assert_desired_datasets(
+            self, self.deployer,
+            local_datasets=[
+                ScenarioMixin.MOUNTED_DISCOVERED_DATASET,
+            ],
+            expected_datasets=[],
+            leases=Leases().acquire(
+                now=datetime.now(tz=UTC),
+                dataset_id=ScenarioMixin.DATASET_ID,
+                node_id=uuid4(),
+            )
+        )
+
+    def test_application_mounted_manifestation(self):
+        """
+        If there is an application with attached volume, there is a
+        corresponding desired dataset that has a state of ``MOUNTED``.
+        """
+        assert_desired_datasets(
+            self, self.deployer,
+            desired_manifestations=[],
+            local_datasets=[
+                ScenarioMixin.MOUNTED_DISCOVERED_DATASET,
+            ],
+            local_applications=[
+                Application(
+                    name=u"myapplication",
+                    image=DockerImage.from_string(u"image"),
+                    volume=AttachedVolume(
+                        manifestation=ScenarioMixin.MANIFESTATION,
+                        mountpoint=FilePath(b"/data")
+                    ),
+                ),
+            ],
+            expected_datasets=[
+                ScenarioMixin.MOUNTED_DESIRED_DATASET,
+            ],
+        )
+
+    @given(
+        expected_size=integers(min_value=0),
+    )
+    def test_leased_manifestation(self, expected_size):
+        """
+        If there is a manifesation on this node and lease for the corresponding
+        volume for this node, then the corresponding desired dataset has a
+        state of ``MOUNTED`` and the associated size corresponds to the
+        discovered dataset.
+        """
+        assert_desired_datasets(
+            self, self.deployer,
+            desired_manifestations=[ScenarioMixin.MANIFESTATION],
+            local_datasets=[
+                ScenarioMixin.MOUNTED_DISCOVERED_DATASET.transform(
+                    ['maximum_size'], expected_size,
+                ),
+            ],
+            expected_datasets=[
+                ScenarioMixin.MOUNTED_DESIRED_DATASET.transform(
+                    ['maximum_size'], expected_size,
+                ),
+            ],
+            leases=Leases().acquire(
+                now=datetime.now(tz=UTC),
+                dataset_id=ScenarioMixin.DATASET_ID,
+                node_id=self.deployer.node_uuid,
+            )
+        )
+
+    def test_deleted_leased_manifestation(self):
+        """
+        If there is a manfestation on this node that is deleted and there is a
```

Comment:
This however seems like it'll result in dataset being recreated, which I don't think we want. So this seems wrong.